mmengine/runner/deepspeed_runner.py

```diff
@@ -0,0 +1,771 @@
+        elif self.cfg.filename is not None:
+            filename_no_ext = osp.splitext(osp.basename(self.cfg.filename))[0]
+            self._experiment_name = f'{filename_no_ext}_{self._timestamp}'
+        else:
+            self._experiment_name = self.timestamp
+        self._log_dir = osp.join(self.work_dir, self.timestamp)
+        mmengine.mkdir_or_exist(self._log_dir)
+        # Used to reset registries location. See :meth:`Registry.build` for
+        # more details.
+        self.default_scope = DefaultScope.get_instance(
+            self._experiment_name, scope_name=default_scope)
+        # Build log processor to format message.
+        log_processor = dict() if log_processor is None else log_processor
+        self.log_processor = self.build_log_processor(log_processor)
+        # Since `get_instance` could return any subclass of ManagerMixin. The
+        # corresponding attribute needs a type hint.
+        self.logger = self.build_logger(log_level=log_level)
+
+        # Collect and log environment information.
+        self._log_env(env_cfg)
+
+        # collect information of all modules registered in the registries
+        registries_info = count_registered_modules(
+            self.work_dir if self.rank == 0 else None, verbose=False)
+        self.logger.debug(registries_info)
+
+        # Build `message_hub` for communication among components.
+        # `message_hub` can store log scalars (loss, learning rate) and
+        # runtime information (iter and epoch). Those components that do not
+        # have access to the runner can get iteration or epoch information
+        # from `message_hub`. For example, models can get the latest created
+        # `message_hub` by
+        # `self.message_hub=MessageHub.get_current_instance()` and then get
+        # current epoch by `cur_epoch = self.message_hub.get_info('epoch')`.
+        # See `MessageHub` and `ManagerMixin` for more details.
+        self.message_hub = self.build_message_hub()
+        # visualizer used for writing log or visualizing all kinds of data
+        self.visualizer = self.build_visualizer(visualizer)
+
+        self._load_from = load_from
+        self._resume = resume
+        # flag to mark whether checkpoint has been loaded or resumed
+        self._has_loaded = False
+
+        # load DeepSpeed configuration file
+        self.ds_config = self.cfg.get('ds_config', None)
+        assert self.ds_config is not None, 'ds_config should be specified.'
+
+        self.check_ds_config(self.ds_config)
+
+        # build a model
+        if isinstance(model, dict) and data_preprocessor is not None:
+            # Merge the data_preprocessor to model config.
+            model.setdefault('data_preprocessor', data_preprocessor)
+        self.model = self.build_model(model)
+
+        # initialize the model weights before wrapping it with deepspeed
+        self._weights_initialized = False
+        self._init_model_weights()
```

Comment:
Model should be initialized before wrapping it with `DeepSpeedEngine`.

Reply:
I couldn't find any documentation yet. But, we found this when we reproduce the performance after training.

Maybe there is the other way to handle this.